File: app/detector.py
import cv2
import numpy as np

def detect_watermark(img: np.ndarray) -> tuple[int, int, int, int]:
    """
    Detect ENCAR watermark bounding box.
    Returns (x1, y1, x2, y2).
    """
    h, w = img.shape[:2]
    
    # Method 1 - fixed zone (last 15% height, 25% width right corner)
    x1_fixed = int(w * 0.75)
    y1_fixed = int(h * 0.85)
    x2_fixed = w
    y2_fixed = h
    
    # Cut crop for the zone
    crop = img[y1_fixed:y2_fixed, x1_fixed:x2_fixed]
    
    # Method 2 - contrast/brightness text detection (optional refine)
    gray = cv2.cvtColor(crop, cv2.COLOR_BGR2GRAY)
    
    # threshold to find bright pixels: white/grey text
    _, thresh = cv2.threshold(gray, 200, 255, cv2.THRESH_BINARY)
    
    # Find connected components
    num_labels, labels, stats, centroids = cv2.connectedComponentsWithStats(thresh, connectivity=8)
    
    # Determine the bounding rect around all detected 'text-like' components
    # Start with empty values
    min_x, min_y, max_x, max_y = w, h, 0, 0
    found_any = False
    
    for i in range(1, num_labels): # skip background 0
        area = stats[i, cv2.CC_STAT_AREA]
        # filter out noise and very big blobs that can't be text
        if 50 < area < 10000:
            found_any = True
            left = stats[i, cv2.CC_STAT_LEFT]
            top = stats[i, cv2.CC_STAT_TOP]
            width = stats[i, cv2.CC_STAT_WIDTH]
            height = stats[i, cv2.CC_STAT_HEIGHT]
            
            min_x = min(min_x, left)
            min_y = min(min_y, top)
            max_x = max(max_x, left + width)
            max_y = max(max_y, top + height)
            
    if found_any:
        # Translate local crop coordinates back to global
        final_x1 = x1_fixed + min_x
        final_y1 = y1_fixed + min_y
        final_x2 = x1_fixed + max_x
        final_y2 = y1_fixed + max_y
        
        # Add 10px padding
        final_x1 = max(0, final_x1 - 10)
        final_y1 = max(0, final_y1 - 10)
        final_x2 = min(w, final_x2 + 10)
        final_y2 = min(h, final_y2 + 10)
        return final_x1, final_y1, final_x2, final_y2
    
    # If refining fails, return the fixed box as fallback fallback
    return x1_fixed, y1_fixed, x2_fixed, y2_fixed

File: src/imgutil/detector.rs
use anyhow::{Context, Result};
use opencv::{
    core,
    imgproc,
    prelude::*,
};

/// Returns bounding boxes (in original image coordinates) of all detected
/// watermark regions. Detection is limited to the bottom-right corner.
pub fn detect_watermark(img: &core::Mat) -> Result<Vec<core::Rect>> {
    let rows = img.rows();
    let cols = img.cols();

    // Focus on bottom-right corner: right 25% × bottom 8%.
    let zone = core::Rect::new(
        (cols as f64 * 0.75) as i32,
        (rows as f64 * 0.92) as i32,
        (cols as f64 * 0.25) as i32,
        (rows as f64 * 0.08) as i32,
    );
    let full_img_rect = core::Rect::new(0, 0, cols, rows);
    let zone = zone & full_img_rect;
    if zone.empty() {
        return Ok(vec![]);
    }

    let crop = core::Mat::roi(img, zone).context("roi crop failed")?;

    // Grayscale (img is in native BGR format).
    let mut gray = core::Mat::default();
    imgproc::cvt_color(&crop, &mut gray, imgproc::COLOR_BGR2GRAY, 0)
        .context("cvt_color BGR→GRAY failed")?;

    // Otsu threshold.
    let mut thresh = core::Mat::default();
    imgproc::threshold(
        &gray,
        &mut thresh,
        0.0,
        255.0,
        imgproc::THRESH_BINARY | imgproc::THRESH_OTSU,
    )
    .context("threshold failed")?;

    // Connected components with stats.
    let mut labels  = core::Mat::default();
    let mut stats   = core::Mat::default();
    let mut centroids = core::Mat::default();
    let num_labels = imgproc::connected_components_with_stats(
        &thresh,
        &mut labels,
        &mut stats,
        &mut centroids,
        8,
        core::CV_32S,
    )
    .context("connected_components_with_stats failed")?;

    let mut results = Vec::new();

    for i in 1..num_labels {
        let area   = *stats.at_2d::<i32>(i, imgproc::CC_STAT_AREA as i32).unwrap_or(&0);
        let left   = *stats.at_2d::<i32>(i, imgproc::CC_STAT_LEFT as i32).unwrap_or(&0);
        let top    = *stats.at_2d::<i32>(i, imgproc::CC_STAT_TOP as i32).unwrap_or(&0);
        let width  = *stats.at_2d::<i32>(i, imgproc::CC_STAT_WIDTH as i32).unwrap_or(&0);
        let height = *stats.at_2d::<i32>(i, imgproc::CC_STAT_HEIGHT as i32).unwrap_or(&0);

        // Ultra-sensitive filter matching Go code.
        if area > 10 && area < 15000 {
            let x1 = (zone.x + left - 3).max(0);
            let y1 = (zone.y + top  - 3).max(0);
            let x2 = (zone.x + left + width  + 3).min(cols);
            let y2 = (zone.y + top  + height + 3).min(rows);
            results.push(core::Rect::new(x1, y1, x2 - x1, y2 - y1));
        }
    }

    Ok(results)
}
